src/Microsoft.AspNetCore.Razor/Parser/HtmlMarkupParser.Block.cs

```diff
@@ -1100,29 +1100,28 @@ private void EndTagBlock(Stack<Tuple<HtmlSymbol, SourceLocation>> tags, bool com
             tags.Clear();
             if (!Context.DesignTimeMode)
             {
-                if (At(HtmlSymbolType.WhiteSpace))
+                var shouldAcceptWhitespaceAndNewLine = true;
+
+                if (Context.LastSpan.Kind == SpanKind.Transition)
                 {
-                    if (Context.LastSpan.Kind == SpanKind.Transition)
-                    {
-                        // Output current span content as markup.
-                        Output(SpanKind.Markup);
+                    var symbols = ReadWhile(
+                        f => (f.Type == HtmlSymbolType.WhiteSpace) || (f.Type == HtmlSymbolType.NewLine));
 
-                        // Accept and mark the whitespace at the end of a <text> tag as code.
-                        AcceptWhile(HtmlSymbolType.WhiteSpace);
-                        Span.ChunkGenerator = new StatementChunkGenerator();
-                        Output(SpanKind.Code);
-                    }
-                    else
+                    if (!At(HtmlSymbolType.OpenAngle))
                     {
-                        AcceptWhile(HtmlSymbolType.WhiteSpace);
+                        // Don't accept whitespace as markup if the end text tag is followed by csharp.
+                        shouldAcceptWhitespaceAndNewLine = false;
                     }
+
+                    PutCurrentBack();
+                    PutBack(symbols);
+                    EnsureCurrent();
                 }
 
-                if (!EndOfFile &&
-                    At(HtmlSymbolType.NewLine) &&
-                    Context.LastSpan.Kind != SpanKind.Transition)
+                if (shouldAcceptWhitespaceAndNewLine)
                 {
-                    AcceptAndMoveNext();
+                    AcceptWhile(HtmlSymbolType.WhiteSpace);
+                    Optional(HtmlSymbolType.NewLine);
```

Comment:
This doesn't line up w/ the `ReadWhile()`. That'll consume multiple intervening lines but this will stop at the end of the current line. If that's intentional, need comments here explaining why.